domain/port/state/state.go

```diff
@@ -0,0 +1,282 @@
+	}
+	return groupedPortRanges, nil
+}
+
+// UpdateUnitPorts opens and closes ports for the endpoints of a given unit.
+//
+// NOTE: This method opens ports first, and then closes ports. However, it is recommended
+// that your service layer reconciles overlapping port ranges before calling this method.
+func (st *State) UpdateUnitPorts(ctx context.Context, unit string, openPorts, closePorts network.GroupedPortRanges) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	unitUUID := unitUUID{UUID: unit}
+
+	endpointsUnderActionSet := set.NewStrings()
+	for endpoint := range openPorts {
+		endpointsUnderActionSet.Add(endpoint)
+	}
+	for endpoint := range closePorts {
+		endpointsUnderActionSet.Add(endpoint)
+	}
+	endpointsUnderAction := endpoints(endpointsUnderActionSet.Values())
+
+	getUnitEndpoints, err := st.Prepare(`
+SELECT &endpoint.*
+FROM unit_endpoint
+WHERE unit_uuid = $unitUUID.unit_uuid
+AND endpoint IN ($endpoints[:])
+`, endpoint{}, unitUUID, endpointsUnderAction)
+	if err != nil {
+		return errors.Annotate(err, "preparing get unit endpoints statement")
+	}
+
+	insertUnitEndpoint, err := st.Prepare("INSERT INTO unit_endpoint (*) VALUES ($unitEndpoint.*)", unitEndpoint{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert unit endpoint statement")
+	}
+
+	getProtocols, err := st.Prepare("SELECT &protocol.* FROM protocol", protocol{})
+	if err != nil {
+		return errors.Annotate(err, "preparing get protocol ID statement")
+	}
+
+	getOpenedPortsOnEndpoints, err := st.Prepare(`
+SELECT &endpointPortRange.*
+FROM port_range
+JOIN protocol ON port_range.protocol_id = protocol.id
+JOIN unit_endpoint ON port_range.unit_endpoint_uuid = unit_endpoint.uuid
+WHERE unit_endpoint.unit_uuid = $unitUUID.unit_uuid
+AND unit_endpoint.endpoint IN ($endpoints[:])
+`, endpointPortRange{}, unitUUID, endpointsUnderAction)
+	if err != nil {
+		return errors.Annotate(err, "preparing get opened ports on endpoints statement")
+	}
+
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert port range statement")
+	}
+
+	clearPortRange, err := st.Prepare("DELETE FROM port_range WHERE unit_endpoint_uuid IN ($endpointUUIDs[:]) ", endpointUUIDs{})
+	if err != nil {
+		return errors.Annotate(err, "preparing clear port range statement")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		// Get the endpoints under action which currently exist in state.
+		// Cache the endpoint's UUIDs for later use.
+		var currenUnitEndpoint []endpoint
+		err = tx.Query(ctx, getUnitEndpoints, unitUUID, endpointsUnderAction).GetAll(&currenUnitEndpoint)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			currenUnitEndpoint = []endpoint{}
+		} else if err != nil {
+			return domain.CoerceError(err)
+		}
+
+		foundEndpoints := set.NewStrings()
+		endpointsUnderActionToUUID := map[string]endpointUUID{}
```

Comment:
I'm not sure I understand the purpose of this map. It seems to only be used to loop through the key/values later on, in which case it could just be a slice

Reply:
I think this is a result of refactors. You're right, we never actually look up a value in the map, just iterate across it

Reply:
Done